home_work/home_work1.py

```diff
@@ -0,0 +1,150 @@
+import json
+from pydantic import BaseModel
+from abc import (
+    ABC,
+    abstractmethod
+)
+from datetime import date
+from contextlib import contextmanager
+
+
+class BookModel(BaseModel):
+    title: str
+    author: str
+    created_date: date
+
+
+class JournalModel(BookModel):
+    last_printed_day: date
+
+
+class AbstractReading(ABC):
+    @abstractmethod
+    def reading_info(self):
+        raise NotImplementedError
+
+
+class Book(AbstractReading):
+    def __init__(self, model: BookModel):
+        self._model = model
+
+    def reading_info(self):
+        return f'Book {self._model.title} writed by {self._model.author} from {self._model.created_date}year'
+
+
+class Journal(Book):
+    def __init__(self, model: JournalModel):
+        self._model = model
+
+
+def add_book_notif(func):
+    def wrapper(self, reading):
+        result = func(self, reading)
+        return f'Book{reading._model.title}was added'
+    return wrapper
+
+
+def check_book(func):
+    def wrapper(self, title):
+        for reading in self.list_of_books:
+            if reading._model.title == title:
+                result = func(self, reading)
+                return f'Book {reading._model.title} deleted'
+    return wrapper
+
+
+class Library:
+    def __init__(self):
+        self.list_of_books: list = []
+
+    @add_book_notif
+    def add_material(self, reading):
+        self.list_of_books.append(reading)
+
+    @check_book
+    def del_material(self, reading):
+        self.list_of_books.remove(reading)
+
+    def books_on_the_library(self):
+        for book in self.list_of_books:
+            print(book.reading_info())
+
+    def books_generator(self, author):
+        for book in self.list_of_books:
+            if book._model.author == author:
+                yield f'Book {book._model.title} writed by {book._model.author}'
+
+    @contextmanager
+    def download_list_manager(self):
+        data = []
+        for book in self.list_of_books:
+            data.append({
+                'title': book._model.title,
+                'author': book._model.author,
+                'created_date': book._model.created_date.isoformat(),
+            })
+        with open('book_list1.json', 'w') as file:
+            json.dump(data, file, indent=4)
+
+    @contextmanager
+    def read_list(self, file):
+        with open(file, 'r') as file_r:
+            book_data = json.load(file_r)
+            for book_info in book_data:
+                title = book_info['title']
+                author = book_info['author']
+                created_date = date.fromisoformat(book_info['created_date'])
```

Comment:
instead of converting field each time, you can add validator for this field and in case of string convert to date, and from date to string. In this case you'll need only to make BookModel(**book_data).
Same for converting from model to json, you can user self._models.model_dump_json().

Reply:
Done, I think